### PR TITLE
Add advancedcontentfilter addon

### PR DESCRIPTION
**Pull request by MrPetovan**
Closes https://github.com/friendica/friendica/issues/4709

Let me explain.

I started working on this addon a couple weeks ago, at first it was just a simple POC, would the Symfony Expression Language (https://github.com/symfony/expression-language) allow Andy H3 to have his complex rule-based content filter system? It turned out that it worked like a charm almost at first try.

Now that the fun part of the work was over, I had to do the gruntwork: forms, lists, CRUD (Create, Read, Update, Delete), errors, and a hefty help page.

Halfway through the add form, I got hit by a severe lack of motivation. I had done this time and again, I didn't have to prove that I could do it, there was no challenge, no tingling to it. And yet it had to be done. How could I avoid to write Yet Another PHP Form?

Lately, I've been considering writing a SPA (Single Page Application) for Friendica, using the API as data provider, but the size of the task made it kind of overwhelming. Why not use the addon as an excuse to test new things, see how it goes?

Since it was a very simple use-case, only a single entity to manage without entity links, I looked for the simplest modern PHP backend and Javascript frontend. I had heard about the Slim framework (https://github.com/slimphp/Slim) through the Chronicle service (https://github.com/paragonie/chronicle), and I settled for VueJS (https://github.com/vuejs/vue) as it wasn't React and it was touted as simpler than AngularJS for apparently the same use.

The Slim framework really has a small footprint when combined with Friendica session, DB handler, authentication, with just under 70 lines of code for routing and error handling.

I was a little less happy with VueJS. The base framework isn't that big, but I had to (?) import vue-resource to perform the API calls, and it definitely felt like a duplication with our existing JQuery library.

The result, however, is beyond anything I could have hoped. Both the HTML for the settings page and the needed custom JS fit into less than 100 lines each. Once I got the hang of VueJS very peculiar way of manipulating the HTML, it became easy to add custom features on top of an example app I found online.

The whole thing feels fast and modern, and yes, it's about 1000 more files than really needed through the magic of Bower/npm package managers, but it's only hard disk, right?

I'm now considering using Slim to break down the massive `include/api.php` file into digestable classes while keeping the routing we kind of have.

Anyway, enjoy the snappiest part of the Friendica interface!

Depends on https://github.com/friendica/friendica/pull/4851

Reviewers: The real deal starts at https://github.com/friendica/friendica-addons/pull/586/commits/b02724f867b1f210d464c8e44d4a3f0d452b9d7f

### CODE REVIEW

File: advancedcontentfilter/doc/advancedcontentfilter.md [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+* ``===`` (identical)
+* ``!=`` (not equal)
+* ``!==`` (not identical)
+* ``<`` (less than)
+* ``>`` (greater than)
+* ``<=`` (less than or equal to)
+* ``>=`` (greater than or equal to)
+* ``matches`` (regex match)
+
+    To test if a string does *not* match a regex, use the logical ``not``
```
Comment (tobiasd) [inline advancedcontentfilter/doc/advancedcontentfilter.md]: Could you please stick to the one-sentence-per-line rule for the docs?

File: advancedcontentfilter/src/middlewares.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+$container = $slim->getContainer();
+
+// Error handler based off https://stackoverflow.com/a/48135009/757392
+$container['errorHandler'] = function () {
+	return function(Psr\Http\Message\RequestInterface $request, Psr\Http\Message\ResponseInterface $response, Exception $exception)
+    {
```
Comment (tobiasd) [inline advancedcontentfilter/src/middlewares.php]: Mixed indentation characters

**Comment by MrPetovan** [comment]
Probably the wrong one.

**Comment by AndyHee** [comment]
I can see the dilemma in terms of the documentation.

At the moment, it's not suitable for users at all. To have a such detailed and complex explantation is useful for those who want to dig deeper. But we need something of more clarity that basically centres around common examples. 

I think, it might be better to have a separate very basic help page in addition to more advanced explanations. TBH, the current help looks formidable, may I say.

I am happy to write his basic explanation that would allow users to copy and paste a rule.   

I checked out both PR and run composer install at the core.
friendica-addons/pull/586
friendica/pull/4851
But for some reason I cannot even "add" anything with the plus sign. Am I not doing this correctly?

**Comment by AndyHee** [comment]
```
Loading failed for the <script> with source “https://friendica.hubup.pro/addon/advancedcontentfilter/vendor/asset/vue/dist/vue.min.js”.
advancedcontentfilter:435
ReferenceError: Vue is not defined
[Learn More]
advancedcontentfilter.js:1:1
```

**Comment by MrPetovan** [comment]
Got it. There's a `.gitignore` file that prevented the add of the `advancedcontentfilter/vendor/asset/vue/dist/` folder to the branch, so the library was never packaged with the addon.

Until I find a workaround, you need to run
```
cd addon/advancedcontentfilter
../../bin/composer.phar install
```
To get the required library.

**Comment by AndyHee** [comment]
So far the composer only returns: "Nothing to install or update"

I'll try again tomorrow.

**Comment by MrPetovan** [comment]
Please delete your vendor folder and try again.

**Comment by AndyHee** [comment]
Got it! The modal form comes up properly now!

More testing tomorrow.

**Comment by AndyHee** [comment]
Next, create table 'friendica.advancedcontentfilter_rules'

**Comment by annando** [comment]
I guess I will wait until @AndyHee finished his testing and @MrPetovan has clarified this vendor thing.

**Comment by MrPetovan** [comment]
My current plan is to move both the `vue` dependency to Friendica core, this is the only way to make it available through a git/composer update. In the process I'll probably drop `vue-resource` to use JQuery instead.

**Comment by MrPetovan** [comment]
This has been done, the file counter is down to a, huh, reasonable number.

**Comment by MrPetovan** [comment]
Turns out I should have done what I suggested @AndyHee and deleted my vendor folder completely, this would have saved me a headache.

**Comment by AndyHee** [comment]
At the moment, I can't actually

**Comment by rabuzarus** [comment]
> Lately, I've been considering writing a SPA (Single Page Application) for Friendica, using the API as data provider

Really cool. I support the idea.

**Comment by MrPetovan** [comment]
@AndyHee Friendica Core adds a new hook in the database structure definition for hooks to add tables. When you enable the addon, a database structure update is performed to add the relevant table to your database. Can you please try to "Reload you Active Addons", or disabling/enabling the advancedcontentfilter addon again? The table should be added to your database during the addon install, unless your Friendica MySQL user can't create tables?

**Comment by AndyHee** [comment]
That sounds easy. I'll try this.

**Comment by MrPetovan** [comment]
Thanks, I guess the timeout was caused by the forced database structure update on addon activation, you may have missed a past update, and the catch up took too long for a browser request.

**Comment by AndyHee** [comment]
I see, thanks.

I noticed on the help page there is some inconsistency in the "Supported variables" section. It says "author-name" for instance but "author_name" appears to be the deployed format for the variables.

**Comment by MrPetovan** [comment]
Indeed, the correct name to use in the expressions is `author_name` because otherwise the hyphen is interpreted as a minus operator. The hyphen comes from the database field name.

**Comment by MrPetovan** [comment]
> I noticed on the help page there is some inconsistency in the "Supported variables" section. It says "author-name" for instance but "author_name" appears to be the deployed format for the variables.

Fixed.

**Comment by AndyHee** [comment]
I somehow still experience high resource consumption when switching branches.

`mysqld 7.2% 99.44 MB mysql `

Could this this be  related to the table being removed?

**Comment by MrPetovan** [comment]
I don't think so, I never experienced the removal of the table when I switched branches. I still have the `advancedcontentfilter_rules` table in my database at the moment even though I don't have the `task/add-advancedcontentfilter-addon` branch active.

Even then, it shouldn't take so long to drop a table with no foreign keys.

**Comment by AndyHee** [comment]
Thanks for shedding light on this.

It must be something else going on with the database, very likely not related to the dropping of the table. I rebooted the server and the resource consumption is still very high.

**Comment by MrPetovan** [comment]
I'm sorry I can't troubleshoot further, how's the addon otherwise?

**Comment by AndyHee** [comment]
Not to worry!

Yes, it worked very well from what I have observed so far. There was no unexpected behaviour so ever in the course of using it. 

I think, this addon will be a welcome feature!

**Comment by MrPetovan** [comment]
I think so too, I'm planning more work to expose more data to the filter as well. The goal for this addon is for it to be able to completely cover the use cases of the other content filter addons (blockem, langfilter, nsfw) as a one-stop content filter shop.

**Comment by AndyHee** [comment]
This sound great!

Imaging this: In the long, users would be able to add filer rules on the fly.  "Filter posts like this" would populated a range input fields, such as the post's assumed language, author's name, key terms, etc. on the bases of which basic rules can added...  OK, just a _vision._

**Comment by MrPetovan** [comment]
This is especially tricky, even GMail doesn't reliably identify the pattern used when using "Create a filter from those emails".

**Comment by AndyHee** [comment]
True. But the easy of such input fields you have in this GMail "Filter messages like this" function is rather persuasive for average users.

Combining this with the power of procmail style filter recipe would cater for both advanced and normal users.

**Comment by MrPetovan** [comment]
Ah, you mean having a "simple mode" where you fill a few form fields and it generates the expression? It is possible.

**Comment by AndyHee** [comment]
Yes, but I guess some fields could be populated quite easily like "author_link".

Also an advantage of such a simple (activation) mode could be that you don't need to go into the settings. Instead you could just bring it up via the post's menu where you star or delete posts.

**Comment by MrPetovan** [comment]
Ah, the dream of making accessible a heavy abstraction...

**Comment by MrPetovan** [comment]
Let's move this forward.

**Comment by MrPetovan** [comment]
Yeah, it's been caused by https://github.com/friendica/friendica/pull/4951

**Pull request merged**